### DISCUSSION TITLE
ENV variables exposed in client

### DISCUSSION THREAD

**Discussion by kevinvugts** [remix-run/remix#4959] [Q&A]
Hello,

I found out today that Remix is refetching all loaders when posting an action to a route. 
I have the following structure.

root.jsx
configurator.jsx
A folder with nested routes for configurator.jsx

I trigger a post request from a nested route inside configurator.jsx and this is refetching all loaders in the hierarchy.  

But the concerning part is, that it is also refetching the ENVS that are defined in the root.jsx as described per docs: https://remix.run/docs/en/v1/guides/envvars#server-environment-variables

```
export async function loader() {
  return json({
    ENV: {
      STRIPE_PUBLIC_KEY: process.env.STRIPE_PUBLIC_KEY,
      FAUNA_DB_URL: process.env.FAUNA_DB_URL,
    },
  });
}

```
To my surprise, this GET request is triggered on the client side after posting the action. Which exposes all the Environment variables!! This doesn't happen on the initial load of the server.

What goes wrong?

![client-exposed-envs](https://user-images.githubusercontent.com/17518874/209481472-e437e102-04f3-447e-b278-17175885b8c2.png)
![refetches](https://user-images.githubusercontent.com/17518874/209481476-7879697e-4dda-47cc-add9-0737585ab0ee.png)

**Comment by sergiodxa** [comment]
Anything you return from the loaders is available to the browser, that's the point of returning data  from the loader.

On a document request the data is available as an inlined JSON script in the HTML, on a revalidation  it will do a fetch request and show up in the Network tab of the browser.

You don't need to return env variables from the loader to use them on the server, so things like your `FAUNA_DB_URL` should probably be used as `process.env.FAUNA_DB_URL` server-side only.

Only return from the loaders the env variables you need to expose to the browser which should be safe to expose to the browser.
